### PR TITLE
Add ToRegex method to WildcardPattern class

### LINKED ISSUE #19992: Add WildcardPattern ToRegex method
### Summary of the new feature / enhancement

As a user I want to be able to convert a user provided PowerShell wildcard pattern into a regex pattern for use in programs that accept regex.

### Proposed technical implementation details (optional)

It looks like there is already a converter for this but just needs to be made public.

https://github.com/PowerShell/PowerShell/blob/60dc724023bbee8a99448961feb0ff1a9c2cf853/src/System.Management.Automation/engine/regex.cs#L79-L86

https://github.com/PowerShell/PowerShell/blob/60dc724023bbee8a99448961feb0ff1a9c2cf853/src/System.Management.Automation/engine/regex.cs#L936-L953

### PR DESCRIPTION
**Pull request by yotsuda**
## PR Summary

Add public `ToRegex()` method to `WildcardPattern` class for converting PowerShell wildcard patterns to regular expressions.

## PR Context

Resolves #19992

PowerShell users often need to convert wildcard patterns to regular expressions for use in programs that accept regex. While the internal implementation already exists, it was not accessible through a public API.

This PR exposes the conversion functionality by adding a public `ToRegex()` method to the `WildcardPattern` class.

### Usage Example

```powershell
$pattern = [System.Management.Automation.WildcardPattern]::new("*.txt")
$regex = $pattern.ToRegex()
$regexObj = [regex]::new($regex)
$regexObj.IsMatch("file.txt")  # True
```

### Conversion Rules

- `*` (asterisk) converts to `.*` in regex (matches zero or more characters)
- `?` (question mark) converts to `.` (matches exactly one character)
- `[abc]`, `[a-z]` (bracket expressions) preserve their regex meaning (character sets and ranges)
- Literal characters are escaped as needed (e.g., `.` → `\.`)
- For backward compatibility, some optimizations are applied (e.g., standalone `*` returns empty string)

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress)
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
- **User-facing changes**
  - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed: 
- **Testing - New and feature**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- [x] New and old tests pass locally

### Implementation

Added a public `ToRegex()` method to the `WildcardPattern` class that:
- Leverages the existing internal `PatternConvertedToRegex` property
- Returns a string (not `Regex` object) for maximum flexibility
- Respects all `WildcardOptions` (IgnoreCase, CultureInvariant, Compiled)
- Includes comprehensive XML documentation with examples

### Files Changed

- `src/System.Management.Automation/engine/regex.cs` - Added public `ToRegex()` method
- `test/powershell/engine/WildcardPattern.Tests.ps1` - Added comprehensive test coverage (13 tests)

## Testing

Added Pester tests covering:
- Basic wildcard conversions (`*`, `?`, `[]`)
- WildcardOptions handling (IgnoreCase, CultureInvariant)
- Consistency with `WildcardPattern.IsMatch()` behavior
- Edge cases (empty pattern, escaped characters, standalone `*`)
- Complex patterns (multiple wildcards, character ranges)

All 13 tests pass locally.

## Additional Notes

- Pure addition - no breaking changes
- Uses existing optimized conversion logic
- Enables interoperability between PowerShell wildcards and regex-based tools

### CODE REVIEW

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            $regexObj = [regex]::new($regex)
+            $regexObj.IsMatch("anything") | Should -BeTrue
+            $regexObj.IsMatch("") | Should -BeTrue
+        }
+
+        It "Handles escaped wildcard characters" {
+            $pattern = [System.Management.Automation.WildcardPattern]::new("file`*.txt")
+            $regex = $pattern.ToRegex()
+            # The asterisk should be escaped in the regex
+            $regex | Should -Match "file.*\.txt"
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: Fixed in 88c891f04

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        It "Converts asterisk wildcard to regex" {
+            $pattern = [System.Management.Automation.WildcardPattern]::new("*.txt")
+            $regex = $pattern.ToRegex()
+            $regex | Should -Match "\.txt"
+        }
+
+        It "Converts question mark wildcard to regex" {
+            $pattern = [System.Management.Automation.WildcardPattern]::new("test?.log")
+            $regex = $pattern.ToRegex()
+            $regex | Should -Match "test.*\.log"
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: Fixed in 88c891f04

File: test/powershell/engine/WildcardPattern.Tests.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: Fixed in 88c891f04

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,129 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+Describe "WildcardPattern.ToRegex Tests" -Tags "CI" {
+    Context "Basic wildcard to regex conversion" {
+        It "Converts asterisk wildcard to regex" {
+            $pattern = [System.Management.Automation.WildcardPattern]::new("*.txt")
+            $regex = $pattern.ToRegex()
+            $regex | Should -Match "\.txt"
```
Comment (iSazonov) [inline test/powershell/engine/WildcardPattern.Tests.ps1]: I'd expect we check exactly expected string in all tests.

Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: @iSazonov Thank you for the feedback. Updated all tests to use -BeExactly with exact expected strings in f2355ed91.

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+        It "Generated regex matches same strings as WildcardPattern.IsMatch" {
+            $testCases = @(
+                @{ Pattern = "*.txt"; TestString = "file.txt"; Expected = $true }
+                @{ Pattern = "*.txt"; TestString = "file.log"; Expected = $false }
+                @{ Pattern = "test?.log"; TestString = "test1.log"; Expected = $true }
+                @{ Pattern = "test?.log"; TestString = "test12.log"; Expected = $false }
+                @{ Pattern = "file[0-9].txt"; TestString = "file5.txt"; Expected = $true }
+                @{ Pattern = "file[0-9].txt"; TestString = "fileA.txt"; Expected = $false }
+            )
```
Comment (iSazonov) [inline test/powershell/engine/WildcardPattern.Tests.ps1]: I wonder why do we need test standard Regex? 
Please remove.

Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: Sorry for including those unnecessary tests. Removed in 7ff9e1685. Added tests for the trailing `.*$` and both-end optimization instead.

File: src/System.Management.Automation/engine/regex.cs [hunk 5]
```diff
@@ -205,6 +205,34 @@ public bool IsMatch(string input)
...
         }
 
+        /// <summary>
+        /// Converts the wildcard pattern to its regular expression equivalent.
+        /// </summary>
+        /// <returns>
+        /// A string that represents the regular expression equivalent of the wildcard pattern.
+        /// </returns>
+        /// <remarks>
+        /// This method converts PowerShell wildcard patterns to regular expression patterns.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/regex.cs]: We should use the same term either "wildcard pattern" or "PowerShell wildcard pattern" everywhere.

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+Describe "WildcardPattern.ToRegex Tests" -Tags "CI" {
+    Context "Basic wildcard to regex conversion" {
```
Comment (iSazonov) [inline test/powershell/engine/WildcardPattern.Tests.ps1]: It seems most of new tests could be in TestCases excluding `Context "Edge cases"`

File: src/System.Management.Automation/engine/regex.cs (hunk 5 above)
Comment (yotsuda) [reply src/System.Management.Automation/engine/regex.cs]: Thank you for the review. Done in 6b861718d.

File: test/powershell/engine/WildcardPattern.Tests.ps1 (hunk 6 above)
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: Thank you for the review. Done in 6b861718d.

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            # but empty regex pattern matches everything
+            $regex | Should -BeExactly ""
+            $regexObj = [regex]::new($regex)
+            $regexObj.IsMatch("anything") | Should -BeTrue
+            $regexObj.IsMatch("") | Should -BeTrue
+        }
+
+        It "Handles escaped wildcard characters" {
+            # Use single quotes to preserve the backtick escape character
+            $pattern = [System.Management.Automation.WildcardPattern]::new('file`*.txt')
```
Comment (iSazonov) [inline test/powershell/engine/WildcardPattern.Tests.ps1]: I'd expect we check all wildcard chars, not only *.

Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: Thanks for the review. I've expanded the test to cover all wildcard characters using TestCases format:

- `*` (asterisk)
- `?` (question mark)
- `[` (opening bracket)
- `]` (closing bracket)

Each test verifies that the escaped wildcard character is converted to the correct regex literal.
Commit: f6e9ddc8c

**Comment by iSazonov** [comment]
@daxian-dbw @SeeminglyScience Looking as the new method could be used should it return `Regex` type instead of `string`? I'd expect users want exactly get and use Regex and they will be forced to convert the string back to Regex. They can always call `ToString()` method themselves  if necessary, for example, for debugging.

File: src/System.Management.Automation/engine/regex.cs [hunk 8]
```diff
@@ -205,6 +205,34 @@ public bool IsMatch(string input)
...
+        /// <example>
+        /// <code>
+        /// var pattern = new WildcardPattern("*.txt");
+        /// string regex = pattern.ToRegex();
+        /// // Returns: "\.txt$"
+        /// </code>
+        /// </example>
+        public string ToRegex()
+        {
+            return PatternConvertedToRegex;
```
Comment (MatejKafka) [inline src/System.Management.Automation/engine/regex.cs]: I'd somewhat expect a method called `ToRegex()` to return a `Regex`, not a string. Is there a good set of flags to set on `Regex` to reliably match what the pattern does? If so, I think it would be cleaner to return a pre-configured `Regex` instance so that the user is guaranteed to get exactly the same behavior.

**Comment by iSazonov** [comment]
@yotsuda I think the method should return Regex. Could you please make the change? Already added tests are good in the case too.

**Comment by yotsuda** [comment]
@iSazonov @MatejKafka Thanks for the feedback!

I've updated `ToRegex()` to return `Regex` instead of `string`. The implementation now caches the `Regex` instance for subsequent calls.

Changes (ce1dcf368):
- `ToRegex()` returns `Regex` directly
- Added `_regex` field to cache the instance
- Removed the internal `PatternConvertedToRegex` property (was only used by `ToRegex()`)
- Added a test to verify that calling `ToRegex()` multiple times returns the same instance

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    It "Converts '<Pattern>' with <OptionName> option" -TestCases @(
+        @{ Pattern = 'TEST'; OptionName = 'IgnoreCase'; Option = [System.Management.Automation.WildcardOptions]::IgnoreCase; Expected = '^TEST$' }
+        @{ Pattern = 'test'; OptionName = 'CultureInvariant'; Option = [System.Management.Automation.WildcardOptions]::CultureInvariant; Expected = '^test$' }
+    ) {
+        param($Pattern, $OptionName, $Option, $Expected)
+        $wildcardPattern = [System.Management.Automation.WildcardPattern]::new($Pattern, $Option)
+        $regex = $wildcardPattern.ToRegex()
+        $regex | Should -BeOfType ([regex])
+        $regex.ToString() | Should -BeExactly $Expected
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1:34]: Added `RegexOptions` verification and `IsMatch` test to confirm the option works correctly.

```powershell
@{ Pattern = 'TEST'; OptionName = 'IgnoreCase'; ...; ExpectedRegexOptions = 'IgnoreCase, Singleline'; TestString = 'test'; ExpectedMatch = $true }
...
$regex.Options.ToString() | Should -BeExactly $ExpectedRegexOptions
$regex.IsMatch($TestString) | Should -Be $ExpectedMatch
```

Note: `RegexOptions.Singleline` is always included as the default in `TranslateWildcardOptionsIntoRegexOptions`.

Fixed in 7a8fbb12d.

[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1:34]: Added test case for the `Compiled` option:

```powershell
@{ Pattern = 'test*'; OptionName = 'Compiled'; ...; ExpectedRegexOptions = 'Compiled, Singleline'; TestString = 'testing'; ExpectedMatch = $true }
```

Fixed in 7a8fbb12d.

File: src/System.Management.Automation/engine/regex.cs [hunk 10]
```diff
@@ -205,6 +196,36 @@ public bool IsMatch(string input)
...
+        /// <item><description>[abc] (bracket expression) converts to [abc] (matches any character in the set)</description></item>
+        /// <item><description>Literal characters are escaped as needed for regex</description></item>
+        /// </list>
+        /// The returned Regex instance is cached for subsequent calls.
+        /// </remarks>
+        /// <example>
+        /// <code>
+        /// var pattern = new WildcardPattern("*.txt");
+        /// Regex regex = pattern.ToRegex();
+        /// // regex.ToString() returns: "^.*\.txt$"
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply src/System.Management.Automation/engine/regex.cs]: Fixed. The `EndWildcardPattern()` method optimizes away the leading `^.*` when a pattern starts with `*`.

```diff
- /// // regex.ToString() returns: "^.*\.txt$"
+ /// // regex.ToString() returns: "\.txt$"
```

Fixed in 7a8fbb12d.

File: test/powershell/engine/WildcardPattern.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            $regex | Should -BeOfType ([regex])
+            $regex.ToString() | Should -BeExactly $Expected
+        }
+
+        It "Returns cached Regex instance on subsequent calls" {
+            $wildcardPattern = [System.Management.Automation.WildcardPattern]::new("*.txt")
+            $regex1 = $wildcardPattern.ToRegex()
+            $regex2 = $wildcardPattern.ToRegex()
+            [object]::ReferenceEquals($regex1, $regex2) | Should -BeTrue
+        }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/engine/WildcardPattern.Tests.ps1]: The current test is still valid. While `ParserOps.NewRegex()` maintains a global cache, the test verifies the observable behavior that repeated `ToRegex()` calls return the same instance. The instance-level `_regex` field caching provides optimization by avoiding repeated global cache lookups.

File: src/System.Management.Automation/engine/regex.cs [hunk 12]
```diff
@@ -63,6 +63,9 @@ public sealed class WildcardPattern
         // we convert a wildcard pattern to a predicate
         private Predicate<string> _isMatch;
 
+        // cached regex for ToRegex() method
+        private Regex _regex;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/regex.cs]: Do we really need the cache? For example `ToWql()` doesn't use cache.
I believe the cache is needed only for multiple retrieves in high performance scenarios, but it is not our case.

Comment (yotsuda) [reply src/System.Management.Automation/engine/regex.cs]: Thanks for the review! I've removed the cache as you suggested. (793b74eea)

**Review by iSazonov — APPROVED** [review]
